apps/emqx_durable_storage/src/emqx_ds_lts.erl

```diff
@@ -229,6 +230,11 @@ trie_copy_learned_paths(OldTrie, NewTrie) ->
 
 %% @doc Lookup the topic key. Create a new one, if not found.
 -spec topic_key(trie(), threshold_fun(), [level()]) -> msg_storage_key().
+topic_key(Trie, ThresholdFun, [<<"$", _/bytes>> | _] = Tokens) ->
+    %% [MQTT-4.7.2-1]
+    %% Put any topic starting with `$' into a separate root so they won't match
+    %% with `#' / `+/...' subscriptions.
+    do_topic_key(Trie, ThresholdFun, 0, ?PREFIX_SPECIAL, Tokens, [], []);
```

Comment:
```suggestion
topic_key(Trie, ThresholdFun, [<<"$", Root/bytes>> | _] = Tokens) ->
    %% [MQTT-4.7.2-1]
    %% Put any topic starting with `$' into a separate root so they won't match
    %% with `#' / `+/...' subscriptions.
    do_topic_key(Trie, ThresholdFun, 0, ?PREFIX_SPECIAL(Root), Tokens, [], []);
```

?

Reply:
...Otherwise topic with different root will be treated the same; it's a bug.

Reply:
Separate roots idea sounds fine. But, just to clarify:

> Otherwise topic with different root will be treated the same; it's a bug.

I don't see why. Currently, we don't throw away the first token here. Are you thinking about situations where this `special` root starts to have `+` edge?

One thing with separate root for each `$token` idea is that thresholds and `threshold_fun` will have subtly different meaning, which complicates understanding a bit.

Reply:
So I haven't managed to catch the bug you've described with 9e67f1f1e054166a91361690db006f2009c3622d, but changed to _multiple separate roots_ approach anyway in 404b198702da513c299e79b9d92b5d54be4697c7.

Reply:
> One thing with separate root for each $token idea is that thresholds and threshold_fun will have subtly different meaning, which complicates understanding a bit.

Using thresholds (optional hints for **optimizing** the storage) to implement features that affect **consistency** is not the right way.